File: common/src/main/java/generations/gg/generations/core/generationscore/common/recipe/GenerationsIngredidents.java
package generations.gg.generations.core.generationscore.common.recipe;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.gson.JsonObject;
import generations.gg.generations.core.generationscore.common.world.recipe.GenerationsIngredient;
import generations.gg.generations.core.generationscore.common.world.recipe.GenerationsIngredientSerializer;
import net.minecraft.network.FriendlyByteBuf;

public class GenerationsIngredidents {
    private static BiMap<String, GenerationsIngredientSerializer<?>> SERIALIZER = HashBiMap.create();

    public static <T extends GenerationsIngredient> void register(String id, GenerationsIngredientSerializer<T> serializer) {
        SERIALIZER.put(id, serializer);
    }

    public static GenerationsIngredientSerializer<?> getSerializer(String id) {
        System.out.println("Trying to get rks ingredient serializer with id " + id);
        System.out.println("The returned value is " + SERIALIZER.getOrDefault(id, null));

        return SERIALIZER.get(id);
    }

    public static GenerationsIngredient fromJson(JsonObject value) {
        var id = value.get("type").getAsString();
        return getSerializer(id).read(value);
    }

    public static JsonObject toJson(GenerationsIngredient value) {
        var obj = new JsonObject();
        obj.addProperty("type", value.getId());
        value.write(obj);
        return obj;
    }

    public static void toNetwork(FriendlyByteBuf buf, GenerationsIngredient ingredient) {
        buf.writeUtf(ingredient.getId());
        ingredient.write(buf);
    }

    public static GenerationsIngredient fromNetwork(FriendlyByteBuf buffer) {
        var id = buffer.readUtf();
        var ingredient = getSerializer(id).read(buffer);

        return ingredient;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/recipe/GenerationsIngredidents.kt
package generations.gg.generations.core.generationscore.common.recipe

import com.mojang.serialization.MapCodec
import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.util.StreamCodecs.asRegistryFriendly
import generations.gg.generations.core.generationscore.common.world.recipe.*
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.ResourceLocation

object GenerationsIngredidents {
//    val REGISTER: Registrar<GenerationsIngredientType<*>> = RegistrarManager.get(GenerationsCore.MOD_ID).builder<GenerationsIngredientType<*>>(GenerationsCore.id("rks_ingredients")).build()
    val MAP = mutableMapOf<ResourceLocation, GenerationsIngredientType<*>>()

//    @JvmField val CODEC = ResourceLocation.CODEC.xmap(REGISTER::get, REGISTER::getId).xmap({ it!! }, { it }).dispatch(GenerationsIngredient::type, GenerationsIngredientType<*>::codec)
//    @JvmField val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, GenerationsIngredient> = ByteBufCodecs.registry(REGISTER.key()).dispatch(GenerationsIngredient::type, GenerationsIngredientType<*>::streamCodec)

    @JvmField val CODEC = ResourceLocation.CODEC.xmap(MAP::get, { it?.id }).xmap({ it!! }, { it }).dispatch(GenerationsIngredient::type, GenerationsIngredientType<*>::codec)
    @JvmField val INGREDIENT_TYPE_STREAM_CODEC = ResourceLocation.STREAM_CODEC.asRegistryFriendly().map(MAP::get, { (it ?: EMPTY).id })
    @JvmField val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, GenerationsIngredient> = INGREDIENT_TYPE_STREAM_CODEC.dispatch({ it.type }, { it?.streamCodec })

    val DAMAGE = register(DamageIngredient.ID, DamageIngredient.CODEC, DamageIngredient.STREAM_CODEC)
    val ITEM = register(ItemIngredient.ID, ItemIngredient.CODEC, ItemIngredient.STREAM_CODEC)
    val ITEM_TAG = register(ItemTagIngredient.ID, ItemTagIngredient.CODEC, ItemTagIngredient.STREAM_CODEC)
    val POKEMON_ITEM = register(PokemonItemIngredient.ID, PokemonItemIngredient.CODEC, PokemonItemIngredient.STREAM_CODEC)
    val TIME_CAPSULE = register(TimeCapsuleIngredient.ID, TimeCapsuleIngredient.CODEC, TimeCapsuleIngredient.STREAM_CODEC)
    val EMPTY = register("empty", MapCodec.unit(EmptyIngredient), StreamCodec.unit(EmptyIngredient));

//    public fun <T> register(name: String, codec: MapCodec<T>, streamCodec: StreamCodec<RegistryFriendlyByteBuf, T>): RegistrySupplier<GenerationsIngredientType<T>> where T: GenerationsIngredient = {
//        REGISTER.register(GenerationsCore.id(name)) { GenerationsIngredientType<T>(codec, streamCodec) }
//    }

    public fun <T> register(name: String, codec: MapCodec<T>, streamCodec: StreamCodec<RegistryFriendlyByteBuf, T>): GenerationsIngredientType<T> where T: GenerationsIngredient {
        val id = GenerationsCore.id(name)
        val type = GenerationsIngredientType<T>(id, codec, streamCodec)

        MAP.put(id, type)
        return type
    }

    @JvmStatic fun init() {
    }
}
